File: Blind75/easy/01-two-pointers/arrays/container-with-most-water.py
"""
11. Container With Most Water
Difficulty: Medium
Pattern: Two Pointers

You are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

Find two lines that together with the x-axis form a container, such that the container contains the most water.

Return the maximum amount of water a container can store.

Example:
Input: height = [1,8,6,2,5,4,8,3,7]
Output: 49

Time: O(n), Space: O(1)
"""

from typing import List


def max_area(height: List[int]) -> int:
    """
    TODO: Implement the solution
    """
    pass


# Test cases
if __name__ == "__main__":
    # Test 1
    height1 = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    print(f"Max Area: {max_area(height1)}")  # 49

    # Test 2
    height2 = [1, 1]
    print(f"Max Area: {max_area(height2)}")  # 1

    # Test 3
    height3 = [4, 3, 2, 1, 4]
    print(f"Max Area: {max_area(height3)}")  # 16

File: Blind75-Practice/medium/02-two-pointers/Java/ContainerWithMostWater.java
/*
 * 11. Container With Most Water
 * Difficulty: Medium
 * Pattern: Two Pointers
 *
 * Problem:
 * You are given an integer array `height` of length `n`. There are `n` vertical lines
 * drawn such that the two endpoints of the `i`th line are `(i, 0)` and `(i, height[i])`.
 *
 * Find two lines that together with the x-axis form a container that holds the most
 * water.
 *
 * Return the maximum amount of water a container can store.
 *
 * Example 1:
 * Input: height = [1,8,6,2,5,4,8,3,7]
 * Output: 49
 *
 * Example 2:
 * Input: height = [1,1]
 * Output: 1
 *
 * Constraints:
 * - n == height.length
 * - 2 <= n <= 10^5
 * - 0 <= height[i] <= 10^4
 *
 * Notes:
 * - The area is controlled by the shorter wall times the width.
 * - Start wide, then move the shorter side inward because moving the taller one cannot
 *   increase the limiting height.
 */

public class ContainerWithMostWater {
    public int maxArea(int[] height) {
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //
    //

    // int left = 0;
    // int right = height.length - 1;
    // int best = 0;
    //
    // while (left < right) {
    //     int width = right - left;
    //     int area = Math.min(height[left], height[right]) * width;
    //     best = Math.max(best, area);
    //
    //     if (height[left] < height[right]) {
    //         left++;
    //     } else {
    //         right--;
    //     }
    // }
    //
    // return best;
        
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        assert sol.maxArea(new int[] {1, 8, 6, 2, 5, 4, 8, 3, 7}) == 49;
        assert sol.maxArea(new int[] {1, 1}) == 1;
        System.out.println("All test cases passed!");
    }
}
